### PR TITLE
logging: Add an option for a custom log header

### PR DESCRIPTION
**Pull request by rbbrns**
There are scenarios when it is necessary to globally redefine a log macro. The existing logging frontend is not sufficient since it works at the function level.

Introduce an option to include a custom application provided header named "zephyr_custom_log.h" at the end of log.h. This allows an application to redefine LOG macros globally.

This change includes a simple test that redefines the core LOG macros to include a custom prefix.

Signed-off-by: Rob Barnes <robbarnes@google.com>

### CODE REVIEW

File: subsys/logging/Kconfig.misc [hunk 1]
```diff
@@ -94,6 +94,15 @@ config LOG_DOMAIN_NAME
 	string "Domain name"
 	default ""
 
+config LOG_CUSTOM_HEADER
```
Comment (yperess) [inline subsys/logging/Kconfig.misc:97]: There's `subsys/logging/Kconfig.mode::CONFIG_LOG_FRONTEND` I think we should:
1. Put this config in the same `Kconfig.mode`
2. Figure out how to document the differences.

It's a shame that this `CONFIG_LOG_FRONTEND` wasn't done at the macro level. Maybe there's a way to change that? I couldn't find one, but maybe someone more familiar with the logging subsystem knows?

File: subsys/logging/Kconfig.misc [hunk 2]
```diff
@@ -94,6 +94,15 @@ config LOG_DOMAIN_NAME
...
+	bool "Include Custom Log Header"
+	help
+		If enabled, a custom application defined header, named "zephyr_custom_log.h",
+		will be included at the end of log.h. This allows for the LOG_* macros to be
+		redefined at the macro level. The application must provide a header named
+		"zephyr_custom_log.h" when this option is enabled. The internal LOG API may
+		change, so use with caution.
```
Comment (yperess) [inline subsys/logging/Kconfig.misc:104]: I would change the last sentence to something like:

The LOG frontend API can be modified by calls to undef and then remapping the entry points to new functionality. This should be done cautiously.

**Comment by yperess** [comment]
@rbbrns we should also mention here that we need this to integrate Zephyr logging with Pigweed's string tokenization which would require us to intercept the log calls at the macro level.

File: subsys/logging/Kconfig.misc (hunk 1 above)
Comment (rbbrns) [reply subsys/logging/Kconfig.misc]: Moved to Konfig.mode. Updated CONFIG_LOG_FRONTEND to mention LOG_CUSTOM_HEADER.

File: subsys/logging/Kconfig.misc (hunk 2 above)
Comment (rbbrns) [reply subsys/logging/Kconfig.misc]: Reworded a bit and added a paragraph to logging/index.rst documentation.

**Comment by rbbrns** [comment]
@yperess Added a comment about pigweed tokenzier to commit message.

**Review by yperess — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**